README.md
# This is not intended as a formal fork

I wanted to work on the [Sarissa](http://dev.abiss.gr/sarissa/) project,  but the original source code is still hosted in a CVS repository (ugh...). Therefore I took the liberty of importing the code to Github. I do not intend to formally fork the project, but having the code here will make it easier for me to create branches and diffs and to share my contributions to the project, which I intend to offer back to the upstream developers.

## How I imported the project

Importing a CVS repository into a Git repositoy proved harder than it should be.

The various answersm at https://stackoverflow.com/questions/881158/is-there-a-migration-tool-from-cvs-to-git turned out to be quite helpful in accomplishing this. I thank the people who contributed the answers there. **However**, it is important to note that the highest rated answer on that page at the time of this writing contains one crucial error: cvs2git requires a full CVS repository to be cloned using rsync. It will **not** work with just a checked out working copy, not even if you manually create an empty CVSROOT folder in it, as is incorrectly stated in that answer. Kudos to StackOverflow user [mhagger](https://stackoverflow.com/users/24478/mhagger) in particular for pointing that out.

# Official project website: [http://dev.abiss.gr/sarissa/](http://dev.abiss.gr/sarissa/)

The original overview from the official project website follows below. (Converted from HTML to GitHub-compatible Markdown language using the excellent conversion tool at [https://domchristie.github.io/to-markdown/](https://domchristie.github.io/to-markdown/). Go check it out!)

## <a name="Overview"></a>Overview

Sarissa is an ECMAScript library acting as a cross-browser wrapper for native XML APIs. It offers various XML related goodies like Document instantiation, XML loading from URLs or strings, XSLT transformations, XPath queries etc and comes especially handy for people doing what is lately known as "AJAX" development.

Supported browsers are Mozilla - Firefox and family, Internet Explorer with MSXML3.0 and up, Konqueror (KDE 3.3+ for sure), Safari and Opera. Konq and Safari offer no XSLT/XPath scripting support AFAIK.

For a HOWTO document that provides examples of common tasks see the [HOWTOs](http://dev.abiss.gr/sarissa/howtos.html) . The API documentation is generated using [JSDoc](http://jsdoctoolkit.org/) and can be found [here](http://dev.abiss.gr/sarissa/jsdoc/index.html) .

The latest version of Sarissa can always be found on [the Sourceforge project page](https://sourceforge.net/projects/sarissa/) . Please send comments, corrections etc through the [mailing list](http://dev.abiss.gr/sarissa/mail-lists.html) or the [forum](https://sourceforge.net/forum/forum.php?forum_id=256492) .

## <a name="License"></a>License

Sarissa is distributed under the [GNU GPL](http://dev.abiss.gr/sarissa/licenses/gpl.txt) version 2 or higher, [GNU LGPL](http://dev.abiss.gr/sarissa/licenses/lgpl.txt) version 2.1 or higher and [Apache Software License](http://dev.abiss.gr/sarissa/licenses/asl.txt) 2.0 or higher. You can use Sarissa according to any of those licenses.

## <a name="Credits"></a>Credits

See [CHANGELOG.txt](http://dev.abiss.gr/sarissa/CHANGELOG.txt) . Please let us know if your name is missing!

Projects used when developing Sarissa:

*   Apache [Maven](http://maven.apache.org/) and [Ant](http://maven.apache.org/) are used to build the Sarissa distributions, documentation and website. Sarissa can nowdays be used as a Maven dependency in your Maven-based webapps, [check it out](/installation.html) .
*   The [Maven JSTools Plugin](http://dev.abiss.gr/mvn-jstools/) is used to produce documentation based on [JSLint](http://www.jslint.com/) and [JSDoc Toolkit](http://jsdoctoolkit.org/) . The JSTools Plugin is also used in Maven based java webapps at runtime, to resolve Sarissa or other JS files and resources.
*   [ECMAUnit](http://kupu.oscom.org/download/) is used for Unit Testing. Check out the Sarissa Unit Tests page [here](/test/testsarissa.html) .
*   The [MobilVox Maven JavaScript Plugin](http://www.mobilvox.com/projects/maven-js-plugin/) is used to create the compressed versions of the Sarissa JS files.
*   The [Syntaxhighlighter](http://code.google.com/p/syntaxhighlighter/) is used for (obviously) syntax highlighting through out the documentation.

Here are some projects using Sarissa, please let us know if you would like to add yours here. This is in the credits simply because if Sarissa was not usefull we would not be working on it. Currently Sarissa has been downloaded about 70k times so we will be maintaining it for some time :-)

*   [XSLT for GWT](http://www.ebessette.com/d/software/XSLTForGWT) uses Sarissa to do exactly what it's name says it does.
*   [Totoe](http://code.google.com/p/totoe/) (Maori for "to split, devide") is a XML parser for GWT which comes with XPath and namespace support.
*   deCarta's [JavaScript API](http://www.decarta.com/products/dds/jsapi.html) , is used to integrate deCarta's Drill Down Server (DDS) geospatial platform into other applications.
*   Oscom's really cool [Kupu WYSIWYG Editor](http://kupu.oscom.org/)
*   [MapBuilder](http://communitymapbuilder.org/) , a modern standards-based web mapping client.
*   [Plone](http://plone.org/) , a user-friendly and powerful open source Content Management System.
*   [Anyterm](http://anyterm.org/) , a web based terminal interface for your servers (it sounds crazy at first but you quickly find yourself wondering "how come i never thought of that?").
*   [Communik8r](http://communik8r.org/) is a new email application for phpGroupWare. communik8r uses AJAX, IMAP sockets and client side XSLT
*   [WXplorer](http://wxplorer.sourceforge.net/) , a very slick, AJAX based file explorer for PHP applications.
*   The [Minesweeper game](http://dev.abiss.gr/sarissa/sample-apps/minesweeper/index.html) featured as a sample application was written by Sean Whalen, who is also the maintainer.
*   Jay Kimble's [JAAJAX Lib for .Net](http://codebetter.com/blogs/jay.kimble/archive/2005/08/16/130777.aspx) .
*   Jason Diamond's [My Ajax.NET library](http://jason.diamond.name/weblog/category/my-ajax-dot-net) .
*   [MojoPortal](http://www.mojoportal.com/) , an Object Oriented web site framework written in C# that runs under ASP.NET on Windows or under Mono on GNU/Linux and Mac OS X.
*   [Taleful.com](http://www.taleful.com/) is an interactive storytelling website. This website was created to give people a place where they can express themselves and share their stories. All are welcome to come, read and post stories, share and collaborate with others in joint story projects, receive creative writing feedback and enhance their imagination.
*   [Freja](http://www.csscripting.com/) (Framework for Restful Javascript Applications) is not yet another Ajax library. It is an Open-Source, MVC, High Level Ajax based Javascript Framework that lets you use your favorite javascript library if you wish. It actually plays well with other javascript toolkits and libraries (prototype, scriptaculous, dojo, etc..).
*   The [HyperScope](http://hyperscope.org/) is a high-performance thought processor that enables you to navigate, view, and link to documents in sophisticated ways. It is a completely client-side system implemented with Dojo and Sarissa.
*   [Ziizo](http://ziizo.com/) , an online service that allows you to create bookmarks, notes, quotations, lyrics, recipes etc. online and access them from anywhere. You can also create a public page and share with the world.
*   The [KSS project](http://kssproject.org/) (formerly known as Kukit) is a javascript framework that aims to allow Ajax development without javascript. It uses stylesheets with CSS-compliant syntax to setup behaviours in the client and a set of well-defined commands that are marshalled back from the server to manipulate the DOM.
*   [MapGuide Open Source](http://mapguide.osgeo.org/) is a web-based platform that enables users to quickly develop and deploy web mapping applications and geospatial web services.
*   [Opportuno](http://www.opportuno.de/) is jobs search engine for employment opportunities in Germany. Its "Instant View" search interface is completely based on the Sarissa library.
*   Dr. Dobb's [FlipBook](http://www.ddjsilverlight.com/flipbook/) , an e-zine built on Silverlight
*   [jstree](http://code.google.com/p/jstree/) , a tree component based on jQuery uses Sarissa to transform XML datasources.
*   [Ajaxterm](http://antony.lesuisse.org/qweb/trac/wiki/AjaxTerm) , a web based terminal (CLI). It was totally inspired and works almost exactly like Anyterm.


CHANGELOG.txt
################################################################################################
                             Changelog and Credits                                            
                                                                                                         
Please help me not forget anyone, send patches with your (or other) 
name and/or change for me to add below. Thanks!
                        
Sarissa is documented with JSDoc [http://jsdoc.sourceforge.net/], tested 
with ECMAUnit [http://kupu.oscom.org/download/]  and built using 
Apache Ant [http://ant.apache.org/], where code is checked using 
some custom code and JSLint [www.JSLint.com] (Cool license, check it out).

Sarissa was inspired by articles and code available on the web, most notably 
the IE Emu Series [http://webfx.eae.net/dhtml/ieemu/] and Cross-Browser DOM 
Wrapper [http://www.webreference.com/programming/javascript/domwrapper/].
 
################################################################################################


0.9.7.8 
    * TODO: fix deprecated IE XSLt emu checks
    * Added Sarissa.VERSION thanks to  Mike Brown 
    * Fixed spelling errors and had drinks thanks to Greg Higgins
    * Had drinks and a good time thanks to Cedric - cedsav
    * Added support for Text Nodes in document.importNode implementation for IE
    * Fixed clearParameters bug in MSXML 3.0 thanks to Hartmut Wynen
    * Enchanced setParameter to allow clearing parameters when null is passed as a param value 
      thanks to Hartmut Wynen
     
0.9.7.7
    * Fixed Sarissa.updateContentFromURI bug (1614011) thanks to Arjen Duursma 

0.9.7.6 
    * Fixed MXXMLWriter warning and XHTML transformToFragment testcase thanks to Waleed K. Abdulla
    * Fixed some typos/errors that broke Safari
    
0.9.7.5 
    * Fixed document.importNode check for IE 7 thanks to Waleed K. Abdulla

0.9.7.4 
    * Emulated XMLDocument via Document for Opera and fixed XPath thanks to Joao Eiras and Michael(tm) Smith
    * Removed MSXML 4/5 progIds (warnings/bugs) thanks to Duncan Booth and Rick Oosterholt
    * Fixed transformToDocument for MSXML 3 (probably the SP2 pooped version) thanks to bug 
      submitter Waleed K. Abdulla and the anonymous poster with the fix [1549749]
    * Fixed transformToFragment bug thanks to Robert Sanders
    * Fixed Sarissa.copyChildNodes thanks to Waleed K. Abdulla
    * Fixed selectSingleNode bug thanks to Joao Eiras 
    * Made document.importNode a bit smarter to stop IE from getting too clever for it's own good
      thanks to Duncan Booth
    * Fixed IE's document.importNode to use document.importNode thanks to Duncan Booth
    * Fixed XSLT document() function for MSXML 6 thanks to Ron MacCracken
    * Added XSLTProcessor.clearParameters for IE

0.9.7.3
    * Fixed transformToFragment bug in IE thanks to Simon Biickert

0.9.7.2
    * Fixed err, typo in sarissa_dhtml.js thanks to aevarga and apb2006 (SF users)
    * Fixed namespace format checking in Sarissa.setXpathNamespaces thanks to Andrew Grooom 

0.9.7.1
    * Added XSLTProcessor.transformToFragment along with ECMAUnit
      tests thanks to John Hax
    * Fixed missing parenthesis typo/error in sarissa_dhtml.js thanks to
      aevarga (SF user) and Marco Ratto
    * Fixed some style.cursor error thanks to aevarga (SF user)
    * Fixed build.xml to convert line endings to be friendly to windows users 
      thanks to John Hax
    * Fixed declarations for window.Sarissa to load properly in different contexts 
      thanks to John Hax
    * Optimised XSLTProcessor.importStylesheet for IE; it will now reload a remote 
      stylesheet from the server only if it contains xsl:include or xsl:import elements
      to resolve those properly, also made it clear in API documentation
    * Added support for importing HTML nodes in document.importNode (only worked with XML)
    * Upgraded JSDoc to 1.9.9.2
    * Added appropriate references to JSLint in credits and documentation (coolest license ever)
    * Fixed  bad CSS font size units from 'pts' to 'pt', guess i was thinking of spare 
      parts or something
    * Changed "gredits" to "credits" in changelog and docos, thanks to Peter (pbb72 SF user). 
      As i  explained i have been unable to ovecome this re-occuring typo for the last 10  
      years due to my coin-op (video game arcades) addiction as a kid ;-)

0.9.7
    * Special thanks to Abiss.gr for sponsoring my new dev machine. Sarissa now is developed
      on an Apple Quad G5 so Safari, Camino and OS X in general can be properly tested
    * Moved all the .load bloat to sarissa_ieemu_load.js and deprecated it. XMLHttpRequest
      objects should be used to load remote files instead
    * Fixed XSLTProcessor's transformToDocument thanks to furtado (sourceforge user)
    * Removed Sarissa.IS_ENABLED_XMLHTTP, check for window.XMLHttpRequest instead
    * Fixed broken-by-typo DOMParser in Safari thanks to Just van den Broecke
    * Fixed Sarissa.getDomDocument inconsistencies between Moz/IE thanks 
      to David Flanagan
    * Added skipCache param in Sarissa.updateContentFromURI thanks to Paul Everitt
    * Fixed Sarissa.getDomDocument in Safari thanks to Paul Everitt
    * Fixed oDoc.load in Safari thanks to Jeroen Ferdinandus 
    * Fixed XPath example errors in documentation thanks to Phil Nash
    * Fixed bug in Sarissa.updateContentFromNode() thanks to Don Hopkins
    * Enchanced documentation to work around a moz bug (xmlhttp.send(null) 
      sends no content-length header) thanks to Mike Branstein
    * Fixed bug where window.Node was overwritten in Moz when sarissa was evaluated
      dynamically (talk about extreme usecases) thanks to John Hax
    * Added callback function parameter to Sarissa.updateContentFromURI and 
      Sarissa.updateContentFromNode thanks to Ron MacCracken and Dave Hall
    * Fixed parseError.errorCode bug and documentation thanks to Bernhard van Woerden

sarissa_ieemu_load.js
/**
 * ====================================================================
 * About
 * ====================================================================
 * Sarissa cross browser XML library - IE .load eulation (deprecated)
 * @version @sarissa.version@
 * @author: Manos Batsis, mailto: mbatsis at users full stop sourceforge full stop net
 *
 * This script emulates Internet Explorer's .load method of DOM Document objects. 
 *
 * All functionality in this file is DEPRECATED, XMLHttpRequest objects 
 * should be used to load XML documents instead 
 *
 * @version @sarissa.version@
 * @author: Manos Batsis, mailto: mbatsis at users full stop sourceforge full stop net
 * ====================================================================
 * Licence
 * ====================================================================
 * Sarissa is free software distributed under the GNU GPL version 2 (see <a href="gpl.txt">gpl.txt</a>) or higher, 
 * GNU LGPL version 2.1 (see <a href="lgpl.txt">lgpl.txt</a>) or higher and Apache Software License 2.0 or higher 
 * (see <a href="asl.txt">asl.txt</a>). This means you can choose one of the three and use that if you like. If 
 * you make modifications under the ASL, i would appreciate it if you submitted those.
 * In case your copy of Sarissa does not include the license texts, you may find
 * them online in various formats at <a href="http://www.gnu.org">http://www.gnu.org</a> and 
 * <a href="http://www.apache.org">http://www.apache.org</a>.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY 
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
 * WARRANTIES OF MERCHANTABILITY,FITNESS FOR A PARTICULAR PURPOSE 
 * AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
if(!_SARISSA_IS_IE){
    if(_SARISSA_HAS_DOM_CREATE_DOCUMENT){
        /**
         * <p>Ensures the document was loaded correctly, otherwise sets the
         * parseError to -1 to indicate something went wrong. Internal use</p>
         * @private
         */
        Sarissa.__handleLoad__ = function(oDoc){
            Sarissa.__setReadyState__(oDoc, 4);
        };
        function SarissaParseError() {
            this.errorCode = 0;
        };
        /**
        * <p>Attached by an event handler to the load event. Internal use.</p>
        * @private
        */
        _sarissa_XMLDocument_onload = function(){
            Sarissa.__handleLoad__(this);
        };
        /**
         * <p>Sets the readyState property of the given DOM Document object.
         * Internal use.</p>
         * @private
         * @argument oDoc the DOM Document object to fire the
         *          readystatechange event
         * @argument iReadyState the number to change the readystate property to
         */
        Sarissa.__setReadyState__ = function(oDoc, iReadyState){
            oDoc.readyState = iReadyState;
            oDoc.readystate = iReadyState;
            if (oDoc.onreadystatechange != null && typeof oDoc.onreadystatechange == "function")
                oDoc.onreadystatechange();
        };
        Sarissa.getDomDocument = function(sUri, sName){
            
            var oDoc = document.implementation.createDocument(sUri?sUri:null, sName?sName:null, null);
            if(!oDoc.onreadystatechange){
            
                /**
                * <p>Emulate IE's onreadystatechange attribute</p>
                */
                oDoc.onreadystatechange = null;
            };
            if(!oDoc.readyState){
                /**
                * <p>Emulates IE's readyState property, which always gives an integer from 0 to 4:</p>
                * <ul><li>1 == LOADING,</li>
                * <li>2 == LOADED,</li>
                * <li>3 == INTERACTIVE,</li>
                * <li>4 == COMPLETED</li></ul>
                */
                oDoc.readyState = 0;
            };
            if(!oDoc.parseError){
                oDoc.parseError = new SarissaParseError();
            };
            oDoc.addEventListener("load", _sarissa_XMLDocument_onload, false);
            return oDoc;
        };
        if(window.XMLDocument){
            /**
            * <p>Keeps a handle to the original load() method. Internal use and only
            * if Mozilla version is lower than 1.4</p>
            * @private
            */
            XMLDocument.prototype._sarissa_load = XMLDocument.prototype.load;
            // NOTE: setting async to false will only work with documents
            // called over HTTP (meaning a server), not the local file system,
            // unless you are using Moz 1.4+.
        
            /**
            * <p>This is deprecated, use XMLHttpRequest to load remote documents instead. 
            * Overrides the original load method to provide synchronous loading for
            * Mozilla versions prior to 1.4 and fix ready state stuff.</p>
            * @deprecated 
            * @returns the DOM Object as it was before the load() call (may be  empty)
            */
            XMLDocument.prototype.load = function(sURI) {
                var oDoc = Sarissa.getDomDocument();
                Sarissa.copyChildNodes(this, oDoc);
                this.parseError.errorCode = 0;
                Sarissa.__setReadyState__(this, 1);
                try {
                    if(this.async == false && _SARISSA_SYNC_NON_IMPLEMENTED) {
                        var tmp = new XMLHttpRequest();
                        tmp.open("GET", sURI, false);
                        tmp.send(null);
                        Sarissa.__setReadyState__(this, 2);
                        Sarissa.copyChildNodes(tmp.responseXML, this);
                        Sarissa.__setReadyState__(this, 3);
                    }else{
                        this._sarissa_load(sURI);
                    };
                }
                catch (objException) {
                        oDoc.parseError.errorCode = -1;
                }
                finally {
                    if (!oDoc.documentElement || oDoc.documentElement.tagName == "parsererror"){
                        oDoc.parseError.errorCode = -1;
                    };
                    if(this.async == false){
                        Sarissa.__handleLoad__(this);
                    };
                };
                return oDoc;
            };
        //if(window.XMLDocument) , now mainly for opera  
        }// TODO: check if the new document has content before trying to copynodes, check  for error handling in DOM 3 LS
        else if(document.implementation && document.implementation.hasFeature && document.implementation.hasFeature('LS', '3.0')){
            Document.prototype.async = true;
            Document.prototype.onreadystatechange = null;
            Document.prototype.load = function(sURI) {
                var oldDoc = Sarissa.getDomDocument();
                Sarissa.copyChildNodes(this, oldDoc, false);
                var parser = document.implementation.createLSParser(this.async ? document.implementation.MODE_ASYNCHRONOUS : document.implementation.MODE_SYNCHRONOUS, null);
                if(this.async){
                    var self = this;
                    parser.addEventListener("load", 
                        function(e) { 
                              self.readyState = 4;
                              Sarissa.copyChildNodes(e.newDocument, self, false);
                              self.onreadystatechange.call(); 
                        }, 
                        false); 
                };
                try {
                    var oDoc = parser.parseURI(sURI);
                    if(!this.async) {
                        Sarissa.copyChildNodes(oDoc, this, false);
                    };
                }
                catch(e){
                    this.parseError.errorCode = -1;
                };
                return oldDoc;
            };
            /**
            * <p>Factory method to obtain a new DOM Document object</p>
            * @argument sUri the namespace of the root node (if any)
            * @argument sUri the local name of the root node (if any)
            * @returns a new DOM Document
            */
            Sarissa.getDomDocument = function(sUri, sName){
                var oDoc = document.implementation.createDocument(sUri?sUri:null, sName?sName:null, null);
                if(!oDoc.parseError){
                    oDoc.parseError = {errorCode:0};
                };
                return oDoc;
            };
        }
        else {
			Sarissa.getDomDocument = function(sUri, sName){
				var oDoc = document.implementation.createDocument(sUri?sUri:null, sName?sName:null, null);
				// looks like safari does not create the root element for some unknown reason
				if(oDoc && (sUri || sName) && !oDoc.documentElement){
				    oDoc.appendChild(oDoc.createElementNS(sUri, sName));
				};
                // attachb to the new object as we have no prototype to use, this is for safari
				if(!oDoc.load) {
        				oDoc.load = function(sUrl) {
        					var oldDoc = document.implementation.createDocument();
        					Sarissa.copyChildNodes(this, oldDoc);
        					this.parseError = {errorCode : 0};
        					Sarissa.__setReadyState__(this, 1);
        					if(this.async == false) {
        						var tmp = new XMLHttpRequest();
        						tmp.open("GET", sUrl, false);
        						tmp.send(null);
        						Sarissa.__setReadyState__(this, 2);
        						Sarissa.copyChildNodes(tmp.responseXML, oDoc);
        						if(!oDoc.documentElement || oDoc.getElementsByTagName("parsererror").length >0){
        					        oDoc.parseError.errorCode = -1;
        					    };
        						Sarissa.__setReadyState__(this, 3);
        						Sarissa.__setReadyState__(this, 4);
        					}
        					else {
        						var xmlhttp = new XMLHttpRequest();
        						xmlhttp.open('GET', sUrl, true);
        						xmlhttp.onreadystatechange = function(){
        						    if (xmlhttp.readyState == 4) {
        							    Sarissa.copyChildNodes(xmlhttp.responseXML, oDoc);
        							    if(!oDoc.documentElement || oDoc.getElementsByTagName("parsererror").length > 0){
        							        oDoc.parseError.errorCode = -1;
        							    };		
        							};
         						Sarissa.__setReadyState__(oDoc, xmlhttp.readyState);
        						};
        						xmlhttp.send(null);
        					};
        					return oldDoc;
        				};
				};
				return oDoc;
			};
        };
    };//if(_SARISSA_HAS_DOM_CREATE_DOCUMENT)
};
//   EOF
